beater/java_attacher/java_attacher.go

```diff
@@ -25,50 +25,369 @@ import (
+	}
+	attacher := JavaAttacher{
+		logger:               logger,
+		enabled:              cfg.Enabled,
+		agentConfigs:         cfg.Config,
+		downloadAgentVersion: cfg.DownloadAgentVersion,
+		rawDiscoveryRules:    cfg.DiscoveryRules,
+		javaBin:              cfg.JavaBin,
+	}
+	for _, flag := range cfg.DiscoveryRules {
+		for name, value := range flag {
+			switch name {
+			case "include-all":
+				attacher.addDiscoveryRule(includeAllRule{})
+			case "include-user":
+				attacher.addDiscoveryRule(userDiscoveryRule{user: value, isIncludeRule: true})
+			case "exclude-user":
+				attacher.addDiscoveryRule(userDiscoveryRule{user: value, isIncludeRule: false})
+			case "include-main":
+				attacher.addCmdLineDiscoveryRule(value, true, "include-main")
+			case "exclude-main":
+				attacher.addCmdLineDiscoveryRule(value, false, "exclude-main")
+			case "include-vmarg":
+				attacher.addCmdLineDiscoveryRule(value, true, "include-vmarg")
+			case "exclude-vmarg":
+				attacher.addCmdLineDiscoveryRule(value, false, "exclude-main")
+			default:
+				logger.Errorf("Unknown discovery rule - '%v'", name)
+			}
+		}
+	}
+	return attacher, nil
+}
+
+func (j *JavaAttacher) addCmdLineDiscoveryRule(regexS string, isIncludeRule bool, argumentName string) {
+	regex, err := regexp.Compile(regexS)
+	if err != nil {
+		j.logger.Errorf("invalid regex for the `%v` argument: %v", argumentName, err)
+		return
+	}
+	j.addDiscoveryRule(cmdLineDiscoveryRule{regex: regex, isIncludeRule: isIncludeRule, argumentName: argumentName})
+}
+
+func (j *JavaAttacher) findFirstMatch(jvm *JvmDetails) discoveryRule {
+	for _, rule := range j.discoveryRules {
+		if rule.match(jvm) {
+			return rule
+		}
+	}
+	return nil
+}
+
+func (j *JavaAttacher) Run(ctx context.Context) error {
+	if !j.enabled {
+		return fmt.Errorf("java attacher is disabled")
+	}
+
+	if runtime.GOOS == "windows" {
+		// in Windows, we run the attacher by a JVM available to the current user in continuous mode and return immediately
+		if err := j.discoverJavaExecutable(); err != nil {
+			return err
+		}
+		return j.attach(ctx, &JvmDetails{})
+	}
+
+	// non-Windows - run discovery and attachment until context is closed
+	for {
+		jvms, err := j.discoverJvmsForAttachment(ctx)
+		if err != nil {
+			j.logger.Infof("error during JVMs discovery: %v", err)
+		} else {
+			if err := j.executeForEachJvm(ctx, jvms, j.attach, true, 30*time.Second); err != nil {
+				j.logger.Error(err)
+			}
+		}
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			time.Sleep(time.Second)
+		}
+	}
+}
+
+func (j *JavaAttacher) discoverJavaExecutable() error {
+	if j.javaBin == "" {
 		if jh := os.Getenv("JAVA_HOME"); jh != "" {
-			cfg.JavaBin = filepath.Join(jh, "/bin/java")
+			j.javaBin = filepath.Join(jh, "/bin/java")
 		} else {
 			bin, err := exec.LookPath("java")
 			if err != nil {
-				return JavaAttacher{}, fmt.Errorf("no java binary found: %v", err)
+				return fmt.Errorf("no java binary found: %v", err)
```

Comment:
**_issue_**: should use the specific modifier to support Go 1.13 error wrapping. See https://go.dev/blog/go1.13-errors#wrapping-errors-with-w

```suggestion
				return fmt.Errorf("no java binary found: %w", err)
```

Reply:
Done here and elsewhere.